### PR TITLE
Backports 0.17.1

### PR DESCRIPTION
**Pull request by haampie**
PR with backports cherrypicked from the develop branch for the 0.17.1 release

Included PRs:
- [x] #25764
- [x] #27846 
- [x] #24675
- [x] #25961
- [x] #27255
- [x] #27472
- [x] #27379
- [x] #27408
- [x] #27676
- [x] #21447
- [x] #27713
- [x] #27707
- [x] #27911
- [x] #26996 
- [x] #27970
- [x] #28031
- [x] #27837 
- [x] #21682
- [x] #27625
- [x] #28037
- [x] #28036
- [x] #14686
- [x] #28072
- [x] #28113

### CODE REVIEW

File: .github/workflows/unit_tests.yaml [hunk 1]
```diff
@@ -199,6 +199,11 @@ jobs:
     runs-on: ubuntu-latest
     container: spack/github-actions:centos6
     steps:
+    - name: Setup git configuration
+      run: |
+        # Need this for the git tests to succeed.
+        git --version
+        . .github/workflows/setup_git.sh
```
Comment (alalazo) [inline .github/workflows/unit_tests.yaml]: Here we are trying to source a file before cloning the Spack repository. I think this line needs to be added twice in each branch of the test (with and without coverage). 

It's unfortunate, but `centos6` is so old that we can't use the `actions/checkout` action inside it if I recall correctly, so we resort to cloning Spack manually.

**Review by alalazo — CHANGES_REQUESTED** [review]

**Comment by alalazo** [comment]
@bonachea We are releasing a patch version of Spack at the end of the week. I have backported 159f4c13f6e6bc57f59bc085cd29a8ce7bea7694 for UPC++. Is that fine with you or are there any more updates needed for the package?

**Comment by bonachea** [comment]
> @bonachea We are releasing a patch version of Spack at the end of the week. I have backported [159f4c1](https://github.com/spack/spack/commit/159f4c13f6e6bc57f59bc085cd29a8ce7bea7694) for UPC++. Is that fine with you or are there any more updates needed for the package?

@alalazo Looks good to me - thanks! 
We don't have any other changes immediately pending.

**Comment by alalazo** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]

**Pull request merged**